src/protocol/primitives/record-batch/record.ts

```diff
@@ -85,3 +111,44 @@ export class Record implements Serializable {
     buffer.writeBuffer(tempBuf);
   }
 }
+
+export class VarIntBytes implements Serializable {
+  constructor(public readonly value: Buffer | null) {}
+
+  public static deserialize(buffer: ReadBuffer): VarIntBytes {
+    const length = VarInt.deserialize(buffer);
+    if (length.value < 0) {
+      return new VarIntBytes(null);
+    }
+
+    const bytes = buffer.readBuffer(length.value);
+
+    return new VarIntBytes(bytes);
+  }
+
+  public serialize(buffer: WriteBuffer): void {
+    if (this.value === null) {
+      new VarInt(-1).serialize(buffer);
+    } else {
+      new VarInt(this.value.length).serialize(buffer);
+      buffer.writeBuffer(this.value);
+    }
+  }
+}
+
+export class VarIntString implements Serializable {
+  constructor(public readonly value: string) {}
+
+  public static deserialize(buffer: ReadBuffer): VarIntString {
+    const varIntBytes = VarIntBytes.deserialize(buffer);
+    if (varIntBytes.value === null) {
+      throw new IllegalArgumentError('VarIntString cannot be null');
```

Comment:
When deserializing headers keys, the Java Kafka clients are throwing `InvalidRecordException("Invalid negative header key size " + headerKeySize);`.

Reply:
Do we need a dedicated exception for that? I will definitely adjust the error message, but at least for now I don't see a value of dedicated error/

Reply:
Is this class used anywhere else?

Reply:
No

Reply:
Then this class can be named `RecordHeaderKey` 😅